### PR TITLE
Improve warnings related to test-only projects

### LINKED ISSUE #4095: Improve TEST-only projects logging for mixed C# / VB .NET solutions
This is a follow-up of validating #4052 and #4063.

For example, for `moq` and `nhibernate`, they have:
- MAIN and TEST projects written in C#
- TEST projects written in VB .NET

And we log the following warnings in the console (and the associated one in the SQ/SC UI).

```
INFO: Sensor C# [csharp]
<... importing results from proto files ... >
INFO: Importing 10 Roslyn reports
INFO: Found 4 MSBuild projects. 2 MAIN projects. 2 TEST projects.
INFO: Sensor C# [csharp] (done) | time=21333ms
INFO: Sensor VB.NET [vbnet]
WARN: This sensor will be skipped, because the current solution contains only TEST files and no MAIN files. Your SonarQube/SonarCloud project will not have results for vbnet files. You can read more about the detection of test projects here: https://github.com/SonarSource/sonar-scanner-msbuild/wiki/Analysis-of-product-projects-vs.-test-projects
INFO: Found 1 MSBuild project. 1 TEST project.
INFO: Sensor VB.NET [vbnet] (done) | time=18ms
```

TODO:
- improve the console logging messages to mention the language more clearly.
```
Before: Found 1 MSBuild project. 1 TEST project.
After: Found 1 MSBuild C# project. 1 TEST project.
```

and also the WARN message
```
WARN: This C# sensor will be skipped, because the current solution contains only TEST files and no MAIN files.
```

- only emit the UI warning if there is no other .NET MAIN project (i.e. the C# sensor should check if there's any MAIN VB.NET code; the VB .NET sensor should check if there's any C# code).

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
Fixes #4095

### CODE REVIEW

**Comment by andrei-epure-sonarsource** [comment]
@costin-zaharia-sonarsource this is the final version

File: its/projects/HtmlCSharpExplicitlyMarkedAsTest/HtmlCSharpExplicitlyMarkedAsTest.csproj [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
```
Comment (costin-zaharia-sonarsource) [inline its/projects/HtmlCSharpExplicitlyMarkedAsTest/HtmlCSharpExplicitlyMarkedAsTest.csproj:4]: Is there any specific reason to have this project on netcore2.2 and not 3.1 as the other one?

File: its/projects/VbMainCsTest/CsTest/UnitTest1.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using NUnit.Framework;
+
+namespace tests
+{
+    public class Tests
+    {
+        [SetUp]
```
Comment (costin-zaharia-sonarsource) [inline its/projects/VbMainCsTest/CsTest/UnitTest1.cs:7]: Is the `SetUp` method needed?

File: its/src/test/java/com/sonar/it/csharp/DoNotAnalyzeTestFilesTest.java [hunk 3]
```diff
@@ -36,35 +36,54 @@
...
 
   @Before
   public void init() {
     TestUtils.reset(ORCHESTRATOR);
   }
 
   @Test
-  public void should_not_increment_test() throws Exception {
-    BuildResult buildResult = Tests.analyzeProjectWithSubProject(temp, PROJECT, "MyLib.Tests", "no_rule");
+  public void with_csharp_only_test_should_not_increment_test() throws Exception {
```
Comment (costin-zaharia-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/DoNotAnalyzeTestFilesTest.java:49]: What about `with_csharp_only_test_should_not_populate_metrics`? increment test seems confusing for me.

File: its/src/test/java/com/sonar/it/csharp/TestProjectTest.java [hunk 4]
```diff
@@ -53,10 +54,19 @@ public void projectIsAnalyzed() {
...
         "Your SonarQube/SonarCloud project will not have results for C# files. " +
         "Read more about how the SonarScanner for .NET detects test projects: https://github.com/SonarSource/sonar-scanner-msbuild/wiki/Analysis-of-product-projects-vs.-test-projects",
-      "Found 1 MSBuild project. 1 TEST project."
+      "Found 1 MSBuild C# project: 1 TEST project."
     );
+    verifyGuiAnalysisWarning(buildResult);
+  }
+
+  // Verifies the analysis warning is raised inside SQ
+  private void verifyGuiAnalysisWarning(BuildResult buildResult) {
```
Comment (costin-zaharia-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/TestProjectTest.java:66]: This method appears 3 times in the project with small variations (at the first look only the language name is different). I suggest to extract and reuse it.

File: its/src/test/java/com/sonar/it/shared/ScannerCliTest.java [hunk 5]
```diff
@@ -56,23 +57,59 @@ public void init() {
   }
 
   @Test
-  public void scannerCliWithRazorPages() {
+  public void scannerCliWithRazorPagesMainCode() {
```
Comment (costin-zaharia-sonarsource) [inline its/src/test/java/com/sonar/it/shared/ScannerCliTest.java:60]: I think we should improve a bit the test name. Just by reading it we can't figure it out what is supposed to test.
WDYT about `givenRazorPagesMainCode_whenScannerForCliIsUsed_displaysWarning` or something similar.

I think we should reconsider a bit also the other names.

File: its/src/test/java/com/sonar/it/shared/ScannerCliTest.java [hunk 6]
```diff
@@ -56,23 +57,59 @@ public void init() {
...
+      .setSourceDirs("main")
+      .setTestDirs("test");
+    BuildResult result = ORCHESTRATOR.executeBuild(scanner);
+
+    assertThat(result.getLogsLines(l -> l.contains("WARN")))
+      .containsExactlyInAnyOrder(
+        "WARN: Your project contains C# files which cannot be analyzed with the scanner you are using. To analyze C# or VB.NET, you must use the SonarScanner for .NET 5.x or higher, see https://redirect.sonarsource.com/doc/install-configure-scanner-msbuild.html"
+      );
+    // The HTML plugin works
+    assertThat(TestUtils.getMeasureAsInt(ORCHESTRATOR, RAZOR_PAGES_PROJECT, "violations")).isEqualTo(2);
```
Comment (costin-zaharia-sonarsource) [inline its/src/test/java/com/sonar/it/shared/ScannerCliTest.java:86]: We analyze `MAIN_AND_TEST_PROJECT` and do the assertion on `RAZOR_PAGES_PROJECT`. Is this intended?

File: its/src/test/java/com/sonar/it/shared/ScannerCliTest.java [hunk 7]
```diff
@@ -56,23 +57,59 @@ public void init() {
...
+  }
+
+  @Test
+  public void scannerCliWithTestHtmlAndCSharp() {
```
Comment (costin-zaharia-sonarsource) [inline its/src/test/java/com/sonar/it/shared/ScannerCliTest.java:90]: The name of the test mentions html and csharp but the project name is `MAIN_AND_TEST_PROJECT`. I think both of them are true but we need to be consistent :)

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/DotNetSensorTest.java [hunk 8]
```diff
@@ -201,22 +243,68 @@ public void whenThereAreNoFiles_logDebug() {
...
     assertThat(logTester.logs(LoggerLevel.WARN)).isEmpty();
     verify(analysisWarnings, never()).addUnique(any());
-    assertThat(logTester.logs(LoggerLevel.INFO)).containsExactly("TEST PROJECTS SUMMARY");
-    assertThat(logTester.logs(LoggerLevel.DEBUG)).hasSize(1);
-    assertThat(logTester.logs(LoggerLevel.DEBUG).get(0)).isEqualTo("No files to analyze. Skip Sensor.");
+    assertThat(logTester.logs(LoggerLevel.DEBUG)).containsExactly("No files to analyze. Skip Sensor.");
+  }
+
+  @Test
+  public void whenThereAreMainFiles_andNotProjects_logToUseScannerForNet() {
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/DotNetSensorTest.java:250]: typo: `NoProjects`

this appears also on the other tests.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM. Just a few minor comments.

File: sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/DotNetSensor.java [hunk 9]
```diff
@@ -115,20 +105,54 @@ private void executeInternal(SensorContext context) {
...
+  /**
+   * If the project does not contain MAIN files OR does not have any found .NET projects (implicitly it has not been scanned with the Scanner for .NET)
+   * we should log a warning to the user, because no files will be analyzed.
+   *
+   * @param hasMainFiles True if MAIN files of this sensor language have been indexed. Can be true only if `anyProjects` is false.
+   * @param anyProjects  True if at least one .NET project has been found in {@link org.sonarsource.dotnet.shared.plugins.FileTypeSensor#execute(SensorContext)}. Can be true only if `hasMainFiles` is false.
+   */
+  private void log(FileSystem fs, boolean hasMainFiles, boolean anyProjects) {
+    boolean hasTestFiles = SensorContextUtils.hasFilesOfType(fs, Type.TEST, pluginMetadata.languageKey());
+    if (anyProjects) {
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/DotNetSensor.java:119]: De Morgan's laws for the win 🎉

**Pull request merged**